<a href="https://colab.research.google.com/github/basawanayya/Own-data-collected-New-House-Construction-Chances-/blob/main/Home_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import io

In [ ]:
from google.colab import files

uploaded = files.upload()

for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))

Saving all collected data.csv to all collected data.csv
User uploaded file "all collected data.csv" with length 6210 bytes


In [ ]:
df= pd.read_csv(io.StringIO(uploaded['all collected data.csv']. decode('utf-8')))

In [ ]:
df.head()

Area name   latitude  longitude  ...  House Type      Type  Construction chances
0  Kusnoor Road  17.312469  76.866931  ...      normal  New home                  high
1  Kusnoor Road  17.307231  76.870097  ...      normal  New home                  high
2  Kusnoor Road  17.311481  76.868011  ...      normal  New home                  high
3  Kusnoor Road  17.307210  76.870207  ...      normal  New home                  high
4   Tilak Nagar  17.315399  76.862952  ...      normal  New home                  high

[5 rows x 12 columns]

In [ ]:
df.dtypes

Area name                            object
latitude                            float64
longitude                           float64
Number of houses                      int64
Total plots                           int64
Area density                        float64
Average age                          object
Material used                        object
No. of old homes(above>25 years)      int64
House Type                           object
Type                                 object
Construction chances                 object
dtype: object

In [ ]:
df.isnull().sum()

Area name                           0
latitude                            0
longitude                           0
Number of houses                    0
Total plots                         0
Area density                        0
Average age                         0
Material used                       0
No. of old homes(above>25 years)    0
House Type                          0
Type                                0
Construction chances                0
dtype: int64

In [ ]:
 from sklearn import preprocessing
encoder = preprocessing.LabelEncoder()

In [ ]:
enc_columns =[]
for column in df.columns:
    if(df[column].dtype == 'object'):
        enc_columns.append(column)
print(enc_columns)

['Area name', 'Average age', 'Material used', 'House Type', 'Type', 'Construction chances']


In [ ]:
for column in enc_columns:
    df[column] = encoder.fit_transform(df[column])

In [ ]:
df.head()

Area name   latitude  longitude  ...  House Type  Type  Construction chances
0          6  17.312469  76.866931  ...           1     0                     0
1          6  17.307231  76.870097  ...           1     0                     0
2          6  17.311481  76.868011  ...           1     0                     0
3          6  17.307210  76.870207  ...           1     0                     0
4          9  17.315399  76.862952  ...           1     0                     0

[5 rows x 12 columns]

In [ ]:
df.dtypes

Area name                             int64
latitude                            float64
longitude                           float64
Number of houses                      int64
Total plots                           int64
Area density                        float64
Average age                           int64
Material used                         int64
No. of old homes(above>25 years)      int64
House Type                            int64
Type                                  int64
Construction chances                  int64
dtype: object

In [ ]:
df['Construction chances'].unique()

array([0, 2, 1])

In [ ]:
# x=df[['area density','Average age','material used','No. of old homes(above>25 years)','House Type','Type']]
x=df.iloc[:,5:11]
y=df['Construction chances']

In [ ]:
#splitting x and y data to test and train
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2)

# Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression 
classifier = LogisticRegression() 
classifier.fit(x_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
#predicting on x_test
pred_log = classifier.predict(x_test)

In [ ]:
from sklearn.metrics import confusion_matrix
#confusion matrix 
cm = confusion_matrix(y_test, pred_log) 
print ("Confusion Matrix : \n", cm)

Confusion Matrix : 
 [[6 0 2]
 [0 1 0]
 [0 0 4]]


In [ ]:
#finding the accuracy score
from sklearn.metrics import accuracy_score
print ("Accuracy : ", accuracy_score(y_test,pred_log))

Accuracy :  0.8461538461538461


In [ ]:
#Create a classification report for the model
from sklearn.metrics import classification_report
print(classification_report(y_test,pred_log))

              precision    recall  f1-score   support

           0       1.00      0.75      0.86         8
           1       1.00      1.00      1.00         1
           2       0.67      1.00      0.80         4

    accuracy                           0.85        13
   macro avg       0.89      0.92      0.89        13
weighted avg       0.90      0.85      0.85        13



# KNN

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
for i in range(1,51,2):
    regressor=KNeighborsClassifier(n_neighbors=i)
    regressor.fit(x_train,y_train)
    pred_knn = regressor.predict(x_test)
    print("Accuracy:",accuracy_score(y_test, pred_knn), "for ",i)

Accuracy: 0.8461538461538461 for  1
Accuracy: 0.7692307692307693 for  3
Accuracy: 0.7692307692307693 for  5
Accuracy: 0.5384615384615384 for  7
Accuracy: 0.7692307692307693 for  9
Accuracy: 0.9230769230769231 for  11
Accuracy: 0.8461538461538461 for  13
Accuracy: 0.8461538461538461 for  15
Accuracy: 0.8461538461538461 for  17
Accuracy: 0.7692307692307693 for  19
Accuracy: 0.7692307692307693 for  21
Accuracy: 0.7692307692307693 for  23
Accuracy: 0.7692307692307693 for  25
Accuracy: 0.38461538461538464 for  27
Accuracy: 0.3076923076923077 for  29
Accuracy: 0.3076923076923077 for  31
Accuracy: 0.3076923076923077 for  33
Accuracy: 0.3076923076923077 for  35
Accuracy: 0.3076923076923077 for  37
Accuracy: 0.3076923076923077 for  39
Accuracy: 0.3076923076923077 for  41
Accuracy: 0.3076923076923077 for  43
Accuracy: 0.3076923076923077 for  45
Accuracy: 0.3076923076923077 for  47
Accuracy: 0.3076923076923077 for  49


# Decision Trees

In [ ]:
#fitting the model
from sklearn.tree import DecisionTreeClassifier

In [ ]:
d_tree = DecisionTreeClassifier()

In [ ]:
d_tree.fit(x_train,y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [ ]:
pred_tree = d_tree.predict(x_test)

In [ ]:
from sklearn.metrics import accuracy_score
print("Accuracy : ",accuracy_score(y_test,pred_tree))

Accuracy :  0.9230769230769231


In [ ]:
from sklearn.metrics import classification_report,confusion_matrix
print(confusion_matrix(y_test,pred_tree))

[[7 0 1]
 [0 1 0]
 [0 0 4]]


In [ ]:
print(classification_report(y_test,pred_tree))

              precision    recall  f1-score   support

           0       1.00      0.88      0.93         8
           1       1.00      1.00      1.00         1
           2       0.80      1.00      0.89         4

    accuracy                           0.92        13
   macro avg       0.93      0.96      0.94        13
weighted avg       0.94      0.92      0.92        13



## **Grid search CV with KNN**

In [ ]:
from sklearn.model_selection import GridSearchCV
param={'n_neighbors':[17]}
est=KNeighborsClassifier()

ht=GridSearchCV(estimator=est, param_grid=param, cv=20)


In [ ]:
ht.fit(x_train,y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_split.py:667: UserWarning: The least populated class in y has only 9 members, which is less than n_splits=20.
  % (min_groups, self.n_splits)), UserWarning)


GridSearchCV(cv=20, error_score=nan,
             estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30,
                                            metric='minkowski',
                                            metric_params=None, n_jobs=None,
                                            n_neighbors=5, p=2,
                                            weights='uniform'),
             iid='deprecated', n_jobs=None, param_grid={'n_neighbors': [17]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

In [ ]:
pred_cv = ht.predict(x_test)

In [ ]:
from sklearn.metrics import accuracy_score
print("Accuracy : ",accuracy_score(y_test, pred_cv))

Accuracy :  0.8461538461538461


Logistic Regression, KNN, Decision Tree and GridSearchCV are used to machine learning modelling. 
The accuracy scores are:

LogisticRegression: 0.84 

KNeighborsClassifier: 0.84

DecisionTreeClassifier: 0.92

GridSearchCV: 0.84 

